pkg/kv/iterators_test.go

```diff
@@ -0,0 +1,377 @@
+package kv_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/go-test/deep"
+	"github.com/treeverse/lakefs/pkg/kv"
+	"github.com/treeverse/lakefs/pkg/kv/kvtest"
+)
+
+const (
+	firstPartitionKey  = "m"
+	secondPartitionKey = "ma"
+	thirdPartitionKey  = "mb"
+)
+
+func TestIterators(t *testing.T) {
+	ctx := context.Background()
+	store := kvtest.GetStore(ctx, t)
+	sm := kv.StoreMessage{
+		Store: store,
+	}
+	m := []kvtest.TestModel{
+		{Name: []byte("a")},
+		{Name: []byte("aa")},
+		{Name: []byte("b")},
+		{Name: []byte("c")},
+		{Name: []byte("d")},
+	}
+
+	// prepare data
+	for i := 0; i < len(m); i++ {
+		err := sm.SetMsg(ctx, firstPartitionKey, m[i].Name, &m[i])
+		if err != nil {
+			t.Fatal("failed to set model", err)
+		}
+
+		err = sm.SetMsg(ctx, secondPartitionKey, m[i].Name, &m[i])
+		if err != nil {
+			t.Fatal("failed to set model", err)
+		}
+	}
+	err := sm.SetMsg(ctx, thirdPartitionKey, []byte("e"), &kvtest.TestModel{})
+	if err != nil {
+		t.Fatal("failed to set model", err)
+	}
+
+	err = sm.SetMsg(ctx, thirdPartitionKey, []byte("f"), &kvtest.TestModel{Name: []byte("notin")})
+	if err != nil {
+		t.Fatal("failed to set model", err)
+	}
+	// first partition - (a,a) (aa,aa) (b,b) (c,c) (d,d)
+	// second partition - (a,a) (aa,aa) (b,b) (c,c) (d,d)
+	// third partition - (e,nil) (f,"notin")
+
+	t.Run("partition iterator listing all values of partition", func(t *testing.T) {
+		itr := kv.NewPartitionIterator(ctx, store, (&kvtest.TestModel{}).ProtoReflect().Type(), firstPartitionKey)
+		if itr == nil {
+			t.Fatalf("failed to create partition iterator")
+		}
+		names := make([]string, 0)
+		for itr.Next() {
+			e := itr.Entry()
+			model, ok := e.Value.(*kvtest.TestModel)
+			if !ok {
+				t.Fatalf("cannot read from store")
+			}
+			names = append(names, string(model.Name))
+		}
+		if itr.Err() != nil {
+			t.Fatalf("unexpected error: %v", itr.Err())
+		}
+		itr.Close()
+
+		if diffs := deep.Equal(names, []string{"a", "aa", "b", "c", "d"}); diffs != nil {
+			t.Fatalf("got wrong list of names: %v", diffs)
+		}
+	})
+
+	t.Run("partition iterator listing values SeekGE", func(t *testing.T) {
+		itr := kv.NewPartitionIterator(ctx, store, (&kvtest.TestModel{}).ProtoReflect().Type(), secondPartitionKey)
+		if itr == nil {
+			t.Fatalf("failed to create partition iterator")
+		}
+		itr.SeekGE([]byte("b"))
+		names := make([]string, 0)
+		for itr.Next() {
+			e := itr.Entry()
+			model, ok := e.Value.(*kvtest.TestModel)
+			if !ok {
+				t.Fatalf("cannot read from store")
+			}
+			names = append(names, string(model.Name))
+		}
+		if itr.Err() != nil {
+			t.Fatalf("unexpected error: %v", itr.Err())
+		}
+		itr.Close()
+
+		if diffs := deep.Equal(names, []string{"b", "c", "d"}); diffs != nil {
+			t.Fatalf("got wrong list of names: %v", diffs)
+		}
+	})
+
+	t.Run("partition iterator failed SeekGE", func(t *testing.T) {
+		itr := kv.NewPartitionIterator(ctx, store, (&kvtest.TestModel{}).ProtoReflect().Type(), "")
```

Comment:
Can we also add `SeekGE` failures when the partition key exists and the key itself is not found?
The 2 failure types behave differently on postgres and dynamodb drivers

Reply:
Not sure I understand the case, if the partition exist and the key doesn't the seek will not fail but return nil.

Reply:
`SeekGE` calls `store.Scan`. Both `postgres/store` and `dynamodb/store` has a `Scan` function that returns a `kv.EntriesIterator` interface. The implementation of this `Scan` calls an `internalScan` function that returns a pointer to a `EntriesIterator` struct. Now, if the failure occurrs in `Scan` (e,g `postgres/store` with empty `partitionKey`) it will return `nil`, but if the failiure is in `InternalScan`, `nil` will be returned to `Scan` which will return am interface pointing to a `nil` object, which is exactly what happend in #4108. Failure due to empty `partitionKey` will not cause the same panic.

Reply:
Got it. To implement it we need a mock for the store. 
It will be implemented on https://github.com/treeverse/lakeFS/pull/4112, not apart of the unit tests